### ISSUE TITLE
SVG Icons not working. Example: Climatology_pi v1.4.14.0

### ISSUE THREAD

**Issue by rgleason** [OpenCPN/plugins#221]
I have checked and made, appropriate changes to icons.h and icons.cpp  (generally done for all plugins just uploaded)
Also new svg icons for  plugin_pi.svg,  plugin_pi_toggled.svg, plugin_pi_rollover.svg were created and put in the data directory.
The path used with getplugindatadir  is intended to point to the plugins data directory.

The icons don't seem to be working, and in place of the toolbar icon we see the green puzzle.

The code repository  is   https://github.com/rgleason/climatology_pi    master branch.

I have similar problems which do not affect operation, with other plugins, such as climatology, tactics, polar, findit, logbook, weather_routing, and watchdog,
A. All plugins have a starting icon in the toolbar except Climatology. Some are not svg.
A. Rollover does not change the toolbar in any of these plugins.
B. Toggled does change in Tactics and Polar, but not the other plugins. Sometimes they don't toggle back.

**Comment by bdbcat** [comment]
Rick...
There are other changes required to support SVG icons.
I think I am a collaborator on climatology?
I'll take a look, hopefully tonight.
Dave

**Comment by bdbcat** [comment]
Rick...
At least two problems:
1.  Simply typo, icons.cpp, line 27:  "climatoloogy"
2.  I see no SVG icons in the "data" directory.

**Comment by rgleason** [comment]
1. I fixed the icons problem. thank you.
2. I created the data dir problem because sean and pavel created a special "data" module that resides on seans repository  https://github.com/rgleason/climatology_pi/tree/33e610855ba7a71b3ed9e614742ba1076043b627

It has "data @ b2f26b6 "  which has all the climiate data and the svg icons!
for some reason I saw there was no "data" dir and tried to create it. Then my data was not accepted to I copied out, then deleted the data dir and copied it back in and then was able to git add data/ 

So now what should I do to get back sean's data dir?

**Comment by rgleason** [comment]
I've tried removing the data directory and adding a new 'virtual' folder
"data @ b2f26b6"  but when I do git status that does not show up to be added.
So that probably is not right.

Maybe the folder just should not exist in local repos?  Then it gets added by github because of the reference to the commit hash?  (That is why I got confused perhaps, there is no placemarker for this type of redirected "data" directory?

**Comment by rgleason** [comment]
Also, it is possible that when building locally, the remote virtual svg icons would not be used?

**Comment by rgleason** [comment]
I tried building locally and using Import Plugin anyway.
I notice now that 
C:\Users\Rick\AppData\Local\opencpn\plugins has no climatology/data directory.

and that the C:\ProgramData\opencpn\plugins\climatology_pi\data directory only has climatology data and not the svg icons.

I closed O and deleted this data directory with the climate data to force an online update.. When I reopened O and tried Climatology it prompted for an internet download as expected and downloaded just the data to 

and that the C:\ProgramData\opencpn\plugins\climatology_pi\data directory only has climatology data and not the svg icons.
but there are no icons.

I am going to try pushing this to cloudsmith next..

**Comment by rgleason** [comment]
I tried msvc tarball from here
https://cloudsmith.io/~opencpn/repos/climatology-beta/packages/?q=1.4.14.1*tarball
it still shows the green puzle and does not change.

I also tried copyiing the svg files intol
C:\ProgramData\opencpn\plugins\climatology_pi\data    
and tested, no dice.

Then copied the svg files to a new climatology//data directory here
C:\Users\Rick\AppData\Local\opencpn\plugins\climatology_pi\data
and tested, no dice still green puzzle.

So I have screwed up the virtual module "data" somehow.
but I wonder why the svg's don't get downloaded with the data and then get used?
in this path?
C:\ProgramData\opencpn\plugins\climatology_pi\data    
but maybe that is not how things work?

Anyway [all the data is here on Sean's pag](https://github.com/rgleason/climatology_pi/tree/33e610855ba7a71b3ed9e614742ba1076043b627)e that gets downloaded.  it is refered to as   
data @ b2f26b6   update data   16 months ago 

https://github.com/seandepagnier/climatology_pi_data/tree/b2f26b6935712665eefd1652e4a70384a645b526

**Comment by rgleason** [comment]
I found the crux of the problem. When I made a data folder, somehow I deleted the gitmodules  entry

https://github.com/rgleason/climatology_pi/commit/72e6101a368268f6df28eb3f72cfd8e6cddb1db6

Now I guess this has to be recreated.

**Comment by rgleason** [comment]
Ok I think I fixed .gitmodules so it has no newline error, but the "data @ b2f26b6"
directory does not appear here as before
https://github.com/rgleason/climatology_pi

So I guess I just have to make a data directory.
.

**Comment by rgleason** [comment]
Well I tried that, and it does not show up in status, I tried pushing it anyway and that did not seem to work.  So now I think I am stuck., do I add  quite literally, "data @  b2f26b6"?

**Comment by rgleason** [comment]
Nope that does not show up in git status either, so there is nothing to add, probably because it is an empty directory.

So I a stuck at this point. Can't go any further.

**Comment by rgleason** [comment]
Maybe I should revert back to v1.4.14.0 which is before I inadvertantlly changed .gitmodules and  data @ hash   and then see what is needed?  (but do this so as not to overwrite history somehow)

https://github.com/rgleason/climatology_pi/commit/f6351e09e1223836b649dfacfcb3a9910be8c4ca

**Comment by rgleason** [comment]
Dave you and Jongough are collabs.

**Comment by rgleason** [comment]
Did this to make a new commit (without changing history) using checkout

Rick@Dart MINGW64 ~/documents/github/climatology_pi (master)
$ git log --oneline
38783b6 (HEAD -> master, origin/master, origin/HEAD) try to fix gitmodules
e5ee164 fix no newline in  gitmodules
b9a0c35 v1.4.14.2
a04acd6 fix screwedup .gitmodules
d57e83f remove local data dir and svg
3f48082 v1.4.14.1 fix icons
72e6101 Try to fix icons
f6351e0 (tag: v1.4.14.0) v1.4.14.0
62ce565 Merge pull request #13 from rgleason/fe2-newest
c0d9239 (tag: v1.4.13.1, origin/fe2-newest) v1.4.13.1 + fe2 v1.0.132 + fix GetPluginDataDir in icons
ca9fb3e Merge pull request #11 from rgleason/jg_fe2-newest
6a540e3 (origin/jg_fe2-newest) Make sure github.com is set
6b41f30 (tag: v1.4.13.0) v1.4.13.0

Using git Checkout seems to be the most reliable way to do this.

    git checkout f6351e0 .  (note the dot is manditory)
    git add -A
    git commit -m "Revert commit: f6351e0"
    git push

----

Rick@Dart MINGW64 ~/documents/github/climatology_pi (master)
$ git checkout f6351e0 .
Updated 5 paths from 789c019

Rick@Dart MINGW64 ~/documents/github/climatology_pi (master)
$ git add -A

Rick@Dart MINGW64 ~/documents/github/climatology_pi (master)
$ git commit -m "Revert commit: f6351e0"
[master a674c40] Revert commit: f6351e0
 5 files changed, 5 insertions(+), 6 deletions(-)
 create mode 160000 data

Rick@Dart MINGW64 ~/documents/github/climatology_pi (master)
$ git push
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 8 threads
Compressing objects: 100% (6/6), done.
Writing objects: 100% (7/7), 635 bytes | 635.00 KiB/s, done.
Total 7 (delta 5), reused 1 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (5/5), completed with 5 local objects.
To https://github.com/rgleason/climatology_pi.git
   38783b6..a674c40  master -> master

**Comment by bdbcat** [comment]
So, is it working now, just without svg?

**Comment by bdbcat** [comment]
I think you need to revert back to something that worked perfectly without SVG icons, and stop there, asking for help.
Dave

**Comment by bdbcat** [comment]
Rick...
Just tried a local build from current github.  Seems to be working, with SVG icons.
What is your status with CircleCI/Cloudsmith build?

**Comment by rgleason** [comment]
Without answering the questions above.
After using checkout to go back to version 1.4.14.0, 
I then fixed my mistake with icon.cpp  spelling of climatology.
Then Pushed that to Cloudsmith as v1.4.14.3  after local testing where the icon remained as a puzzle.
I downloaded the metadata file for climatology 1.4.14.3  msvc and pasted that over my current ocpn-plugin.xml  (the  older climatology msvc v1.4.14.? xml record).

Then I removed everything about climatology f(dll's, climatology directories, data from 
C:\ProgramData\opencpn\plugins
and  C:\Users\Rick\AppData\Local\opencpn\plugins\climatology_pi\data  (removed data and climatology)
also any climatology_pi.dll

Also made sure to remove all climatology tarballs from 
C:\ProgramData\opencpn\plugins\cache\tarballs

Then I started up Opencpn and selected Climatology v1.4.14.3 which showed in the Plugin List.
When I started the plugin, it promped for a download. The download completed with a message. Then I had to select the brown icon to install the plugin. Then with the plugin enabled I was prompted again to download climate data, and it downloaded (I believe to  C:\Users\Rick\AppData\Local\opencpn\plugins\climatology_pi\data  and I believe at the same time the svg icons were downloaded.

Later I found that the climatology files were copied to the normal program data location for windows
C:\ProgramData\opencpn\plugins

After this I noticed that the toolbar icon worked. 
When I click on it the icon changes and the menu comes up.

However when I rollover it with my mouse it does not change to the rollover icon!!!  WHY?

In answer, to Dave's questions, this build in github and cloudsmith work properly:

- Github  v1.4.14.3 fix icons   https://github.com/rgleason/climatology_pi/commit/c30eb3152472a70dda35187fef81912f036d994f

- Cloudsmith  https://cloudsmith.io/~opencpn/repos/climatology-beta/packages/?q=1.4.14.3*tarball

It is bugging me that rollover does not work!

**Comment by rgleason** [comment]
Why doesn't rollover work properly?

**Comment by rgleason** [comment]
I''ve tried adding messages to each icon when it is loaded, but they don't seem to show.
I am now really convinced that the green puzzle icon will always be used in the toolbar when a plugin is installed from a locally created tarball using "Import Plugin". 

There is no way around this other than to push a new version increment to git and download the appropriate metadata file from the climatology  Cloudsmith directory, then paste it into  C:\ProgramData\opencpn\plugins\ocpn-plugins.xml in over the appropriate record, save the file and then open up O and load the plugin. Then the icon will be loaded.  BTW, the svg icons are found and loaded before the climatology data prompt to download.

The problem is there is no use of the "climatology_pi_rollover.svg" icon at the moment that the word "Climatology" pops up!  See attached.

![Climatology_Rollover](https://user-images.githubusercontent.com/2363959/89695894-b2f81500-d8e3-11ea-8343-6a80cb801c88.png)

**Comment by bdbcat** [comment]
Rick...
1. Toolbar rollover has been disabled intentionally, since OCPN 5.0.  You will observe that no tool, either for system or plugin, responds to rollover.
2. "I am now really convinced that the green puzzle icon will always be used in the toolbar when a plugin is installed from a locally created tarball using "Import Plugin". 
Do you observe this for all plugins, or only climatology?

Dave

**Comment by rgleason** [comment]
![climatology_pi_rollover](https://user-images.githubusercontent.com/2363959/89696357-847b3980-d8e5-11ea-801d-1db2f2ad2477.png)

**Comment by rgleason** [comment]
OK I just went through about 10 hours of screwing around, trying to figure out why rollover does not work!!

**Comment by rgleason** [comment]
NOTICE:  ROLLOVER ICONS have been disabled since Opencpn v5.0

**Comment by rgleason** [comment]
Climatology is not a typical case for icons because there is a dedicated separate repository (sean's) for the data with is turned into a module through the use of .gitmodule and a path to the separate repository. It's kind of a virtual repository for "data"

The svg icons reside in that virtual "data" directory.
The other key files are icon.h and icon.cpp where if statements need to be added pointing to the svg icons and naming them.

When installed, the data directory is installed to the new plugin location using getplugindatadir.
For windows that is %localappdata%/opencpn/plugins/climatology/data

**Comment by rgleason** [comment]
"I am now really convinced that the green puzzle icon will always be used in the toolbar when a plugin is installed from a locally created tarball using "Import Plugin".
Do you observe this for all plugins, or only climatology?

I will find out soon.

**Comment by rgleason** [comment]
Dave, I've got pathway issues I can't solve, with LogbookKonni
https://github.com/ptulp/LogbookKonni_pi/issues/24

**Comment by bdbcat** [comment]
Yes, climatology is different in handling of "bundled: data.
Best solution would be create a new  directory "icons", included in the base github repo.  This will require some code changes in the plugin in order to implement.  Or, just live with the fact the the icon will not appear until the plugin is installed and stabilized.

**Comment by rgleason** [comment]
That would not be too hard to do. Maybe I'll try.

**Comment by rgleason** [comment]
Here is what I did.
https://github.com/rgleason/climatology_pi/commit/8cb5614cba246980e0e96ff985877de35715ca46

When I build this locally and us Import Plugins, I still get the green icon, even when shutting down opencpn and restarting.

The problem is that "icons" is not being copied into 
C:\Users\Rick\AppData\Local\opencpn\plugins\climatology_pi

**Comment by rgleason** [comment]
Can't find where the "data" file is copied. It may be built into the Plugin Installer on the Opencpn side.
Searching all climatology files for getplugindatadir I find that only in icons.cpp.
Searching all climatology files for GetpSharedDataLocation()  
I find that climatology_pi.cpp has this at Line 62!!

```
wxString ClimatologyDataDirectory()
{
    wxString s =wxFileName::GetPathSeparator();
    return *GetpSharedDataLocation() + "plugins"
        + s + "climatology_pi" + s + "data" + s;
}

wxString ClimatologyUserDataDirectory()
{
    wxString s = wxFileName::GetPathSeparator();
    return *GetpPrivateApplicationDataLocation() + s + "plugins"
        + s + "climatology_pi" + s + "data" + s;
}
```

So this needs to be fixed too and icons added.

**Comment by rgleason** [comment]
```
wxString ClimatologyDataDirectory()
{
//    wxString s =wxFileName::GetPathSeparator();
//    return *GetpSharedDataLocation() + "plugins"
//        + s + "climatology_pi" + s + "data" + s;

    wxString s =wxFileName::GetPathSeparator();
    return *GetPluginDataDir("climatology_pi") + s + "data" + s;

}



wxString ClimatologyUserDataDirectory()
{
    wxString s = wxFileName::GetPathSeparator();
    return *GetpPrivateApplicationDataLocation() + s + "plugins"
        + s + "climatology_pi" + s + "data" + s;
}

```

What is ClimatologyUserDataDirectory()  used for?    GetPrivateApplicationDataLocation() ? Is this to access the virtual directory on Sean's github?

**Comment by rgleason** [comment]
```
wxString ClimatologyDataDirectory()
{
//    wxString s =wxFileName::GetPathSeparator();
//    return *GetpSharedDataLocation() + "plugins"
//        + s + "climatology_pi" + s + "data" + s;

    wxString s =wxFileName::GetPathSeparator();
    return GetPluginDataDir("climatology_pi") + s + "data" + s;
}
```
Removed the asterick at the beginning and it built and works, but the green icon still shows.  I still have not figured out how to copy the icons folder like the data directory is copied.

I am going to have to figure out where ClimatologyDataDirectory()  is used and make a new wxstring for icons  so it would be something like this.

wxString ClimatologyDataDirectory()
{
    wxString s =wxFileName::GetPathSeparator();
    return GetPluginDataDir("climatology_pi") + s + "icon" + s;
}

But I've got to figure out where the string is used to actually execute the copy.

**Comment by rgleason** [comment]
ClimatologyDataDirectory() is used extensively in ClimateOverlayFactory and CllimatollogyConfigDialog.cpp to copy different types of data.

This is getting to complicated

**Comment by rgleason** [comment]
I need some code to copy the "icons" folder 

Copy icons folder to ClimatologyIconsDirectory()
or copy all the files in the icons folder to  ClimatologyIconsDirectory()

**Comment by bdbcat** [comment]
Rick...
I have a patch, but for some reason the invitation to be a collaborator on this project was rejected. Please issue another invitation.
Thanks
Dave

**Comment by rgleason** [comment]
Thank you. I canceled the invite, then made a new one so it should have been sent to you again I hope. It says "1 invitation" here under Settings > Manage Access

**Comment by rgleason** [comment]
I think I am going to use your patch as an example to make findit_pi  icons work because there is no data directory in findit source, and I left the svg icons in src folder, so they are not available when the plugin is installed. I'll just put them in a new "icons" folder and make the changes needed, following what you've done in climatology.

**Comment by bdbcat** [comment]
Rick...
Pushed my patch successfully.
Let me know how it works
Dave

**Comment by rgleason** [comment]
Dave,
I downloaded the cloudsmith 1.4.14.5  msvc tarball and installed it using "Import Plugin" and it works very nicely!
I like the use of pluginconfigure.cmake for your code.  I am going to see if Jon will include this in frontend2.
![Icons-Directory-Copied-and-Working](https://user-images.githubusercontent.com/2363959/89722192-b14a5200-d9b4-11ea-8697-38b1b03be484.png)

**Comment by rgleason** [comment]
Will use this to fix Findit next!

**Comment by rgleason** [comment]
Have made new svg icons for Findit. Then found that findit_pi.cpp was missing critical #ifdif PLUGINS_USE_SVG statements. I tried to follow climatology, learned some things, but it is still not working.  The original file is on the repos as findit_pi-original.cpp. The plugin works, but the svg icons show as green icon.

Github Repos "master" branch builds but svg icons don't work:  https://github.com/rgleason/FindIt_pi
Tarballs:  https://cloudsmith.io/~opencpn/repos/findit-beta/packages/?q=1.2.5.7*tarball

Changes to findit_pi.cpp

Dumb,  just noticed the icons are not using "_" .   Will see if that makes a difference.
(just

**Comment by rgleason** [comment]
Well, that fixed it, but now I need to get the icons centered.

**Comment by rgleason** [comment]
The tactics svg icon when renamed, shows in the toolbar. The inkscape icons I've made do not fit properly.
These svg icons have 0,0 at the upper right corner just like the tactics_pi svg, and they are 30px x 30px. They are using A4 paper set to transparent, the ICON object (and sub object), is the only object in the list any other layers have been removed. 
Just a white square shows.  However I can load both icons in chrome, edge, firefox, gimp and they appear.

The four icons are below, basically it is just two different icons.
[pi-findit.zip](https://github.com/OpenCPN/plugins/files/5048635/pi-findit.zip)

![findit-Icon](https://user-images.githubusercontent.com/2363959/89744592-71e73880-da7c-11ea-959f-cb8119a38aa5.png)

The Cloudsmith findit-prod repos should have the built versions
https://cloudsmith.io/~opencpn/repos/findit-prod/packages/?q=1.2.6.0*tarball

The github repos is  master branch  https://github.com/rgleason/FindIt_pi

Could use a little guidance here.
